front50-s3/src/main/java/com/netflix/spinnaker/front50/model/TemporarySQSQueue.java

```diff
@@ -0,0 +1,165 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.front50.model;
+
+import com.amazonaws.auth.policy.Condition;
+import com.amazonaws.auth.policy.Policy;
+import com.amazonaws.auth.policy.Principal;
+import com.amazonaws.auth.policy.Resource;
+import com.amazonaws.auth.policy.Statement;
+import com.amazonaws.auth.policy.actions.SQSActions;
+import com.amazonaws.services.sns.AmazonSNS;
+import com.amazonaws.services.sns.model.ListTopicsResult;
+import com.amazonaws.services.sns.model.Topic;
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.model.CreateQueueRequest;
+import com.amazonaws.services.sqs.model.Message;
+import com.amazonaws.services.sqs.model.ReceiptHandleIsInvalidException;
+import com.amazonaws.services.sqs.model.ReceiveMessageRequest;
+import com.amazonaws.services.sqs.model.ReceiveMessageResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PreDestroy;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Encapsulates the lifecycle of a temporary queue.
+ *
+ * Upon construction, an SQS queue will be created and subscribed to the specified SNS topic.
+ * Upon destruction, both the SQS queue and SNS subscription will be removed.
+ */
+public class TemporarySQSQueue {
+  private final Logger log = LoggerFactory.getLogger(TemporarySQSQueue.class);
+
+  private final AmazonSQS amazonSQS;
+  private final AmazonSNS amazonSNS;
+
+  private final TemporaryQueue temporaryQueue;
+
+  public TemporarySQSQueue(AmazonSQS amazonSQS, AmazonSNS amazonSNS, String snsTopicName, String instanceId) {
+    this.amazonSQS = amazonSQS;
+    this.amazonSNS = amazonSNS;
+
+    String snsTopicArn = getSnsTopicArn(amazonSNS, snsTopicName);
+    String sqsQueueName = snsTopicName + "__" + instanceId;
+    String sqsQueueArn = snsTopicArn.substring(0, snsTopicArn.lastIndexOf(":") + 1).replace("sns", "sqs") + sqsQueueName;
+
+    this.temporaryQueue = createQueue(snsTopicArn, sqsQueueArn, sqsQueueName);
+  }
+
+  List<Message> fetchMessages() {
+    ReceiveMessageResult receiveMessageResult = amazonSQS.receiveMessage(
+      new ReceiveMessageRequest(temporaryQueue.sqsQueueUrl)
+        .withMaxNumberOfMessages(10)
+        .withWaitTimeSeconds(5)
+    );
+
+    return receiveMessageResult.getMessages();
+  }
+
+  void markMessageAsHandled(String receiptHandle) {
+    try {
+      amazonSQS.deleteMessage(temporaryQueue.sqsQueueUrl, receiptHandle);
+    } catch (ReceiptHandleIsInvalidException e) {
+      log.warn("Error deleting message, reason: {} (receiptHandle: {})", e.getMessage(), receiptHandle);
+    }
+  }
+
+  @PreDestroy
+  void shutdown() {
```

Comment:
Makes a best attempt to cleanup the queue and subscription.

We'll probably need to do something to detect orphaned queues and remove them periodically.

Reply:
should be fairly easy to do with the instanceid in the queue name